packages/@glimmer/runtime/lib/compiled/opcodes/component.ts

```diff
@@ -174,15 +180,24 @@ APPEND_OPCODES.add(Op.ResolveDynamicComponent, (vm, { op1: _meta }) => {
     let resolvedDefinition = resolveComponent(vm.runtime.resolver, component, meta);
 
     definition = expect(resolvedDefinition, `Could not find a component named "${component}"`);
-  } else if (isCurriedComponentDefinition(component)) {
-    definition = component;
   } else {
-    throw unreachable();
+    definition = component;
   }
 
   stack.pushJs(definition);
 });
 
+APPEND_OPCODES.add(Op.ResolveCurriedComponent, (vm) => {
+  let stack = vm.stack;
+  let definition = check(
+    valueForRef(check(stack.popJs(), CheckReference)),
+    CheckCurriedComponentDefinition
+  );
```

Comment:
This needs to be done in DEBUG mode and actually throw an error that is **not** stripped during publishing. Ember and Glimmer.js users must be able to see this error.

```suggestion
  let stack = vm.stack;
  let value = valueForRef(check(stack.popJs(), CheckReference));
  if (DEBUG && !isCurriedComponentDefinition(value)) {
    throw new Error(`A curried component definition. You may have accidentally done <this.dynamicComponent>, where dynamicComponent was a string instead of a curried component definition. You must use the {{component}} helper to create a component definition when invoking dynamically.`);
  }
  
  let definition = value as CurriedComponentDefinition;
```